### PR TITLE
Add support for incrementally persisting the active block in the peers bootstrapper as a snapshot

### PR DESCRIPTION
**Pull request by richardartoul**
Addresses #900 and depends on #894

### CODE REVIEW

File: src/dbnode/storage/bootstrap/types.go [hunk 1]
```diff
@@ -90,13 +100,11 @@ type ProcessOptions interface {
...
-	// that saves intermediate results to durable storage or not.
-	SetIncremental(value bool) RunOptions
+	// SetIncrementalConfig sets incremental configuration for this bootstrap.
+	SetIncrementalConfig(value IncrementalConfig) RunOptions
 
-	// Incremental returns whether this bootstrap should be an incremental
-	// that saves intermediate results to durable storage or not.
-	Incremental() bool
+	// IncrementalConfig returns the incremental configuration for this bootstrap.
+	IncrementalConfig() IncrementalConfig
```
Comment (prateek) [inline src/dbnode/storage/bootstrap/types.go]: maybe time to rename this guy to `PersistConfig` too (talked to @robskillington, that's the best we could come up with)

File: src/dbnode/storage/bootstrap/types.go [hunk 2]
```diff
@@ -90,13 +100,11 @@ type ProcessOptions interface {
 
 // RunOptions is a set of options for a bootstrap run.
 type RunOptions interface {
-	// SetIncremental sets whether this bootstrap should be an incremental
-	// that saves intermediate results to durable storage or not.
-	SetIncremental(value bool) RunOptions
+	// SetIncrementalConfig sets incremental configuration for this bootstrap.
```
[the remark this answers is not in the document]
Comment (richardartoul) [reply src/dbnode/storage/bootstrap/types.go]: I can add the validate from the first comment but I don't really understand the second comment.

File: src/dbnode/storage/bootstrap/run_options.go [hunk 3]
```diff
@@ -22,35 +22,37 @@ package bootstrap
 
 import "github.com/m3db/m3/src/dbnode/topology"
 
-const (
-	// defaultIncremental declares the intent to by default not perform an
-	// incremental bootstrap.
-	defaultIncremental = false
+var (
+	// defaultPersistConfig declares the intent to by default to perform
+	// a bootstrap without persistence enabled.
```
Comment (prateek) [inline src/dbnode/storage/bootstrap/run_options.go]: hm you say without here but have `Enabled: true` below?

File: src/dbnode/integration/cluster_add_one_node_test.go [hunk 4]
```diff
@@ -258,4 +262,13 @@ func TestClusterAddOneNode(t *testing.T) {
 	for i := range setups {
 		verifySeriesMaps(t, setups[i], namesp.ID(), expectedSeriesMaps[i])
 	}
+
+	if verifyCommitlogCanBootstrapAfterNodeJoin {
+		// Verify that the node that joined the cluster can immediately bootstrap
+		// the data it streamed from its peers from the commit log as soon as
+		// the bootstrapping process completes.
+		require.NoError(t, setups[1].stopServer())
+		startServerWithNewInspection(t, opts, setups[1])
```
Comment (prateek) [inline src/dbnode/integration/cluster_add_one_node_test.go]: do you need to reset the bootstrappers here - i.e. ensure it doesn't use peers again?

File: src/dbnode/storage/bootstrap/bootstrapper/fs/source.go [hunk 5]
```diff
@@ -1155,6 +1157,11 @@ func (s *fileSystemSource) bootstrapFromIndexPersistedBlocks(
 	return res, nil
 }
 
+func (s *fileSystemSource) shouldPersist(runOpts bootstrap.RunOptions) bool {
+	persistConfig := runOpts.PersistConfig()
+	return persistConfig.Enabled && persistConfig.FileSetType == persist.FileSetFlushType
```
Comment (prateek) [inline src/dbnode/storage/bootstrap/bootstrapper/fs/source.go]: +1

File: src/dbnode/integration/cluster_add_one_node_test.go (hunk 4 above)
Comment (richardartoul) [reply src/dbnode/integration/cluster_add_one_node_test.go]: nah startServerWithNewInspection completely reconstructs the bootstrappers

**Pull request merged**